Create datatype and namespace parsers from code created by Cecille

j-ororke opened the pull request:
**Creating namespace parser with data provided by Cecille:**
- Created new TestSpecParsingNamespace.py unit test to test namespaces
- Updated tests.yaml to include TestSpecParsingNamespace.py
- Updated test_metadata.yaml to include TestSpecParsingNamespace.py and reasoning
- Updated matter_testing support module to include new NamespacePathLocation function for Namespace file locations
- Updated spec_parsing to include new functions build_xml_namespaces and parse_namespace, as well as XmlNamespaces and XmlTags dataclasses

**Creating Datatype parser:**
- **spec_parsing module changes:**
- Enhanced XmlDataTypeComponent with fields for docs, types, and constraints
- Added support for detecting optional and nullable fields in XML
- Improved constraint parsing for min/max values and attribute references, basically used for dataTypes
- Improved some error reporting with specific diagnostic messages when parsing dataTypes
- **TestSpecParsingDataType unit test module:**
- Added comprehensive test coverage using Matter 1.4.1 cluster XML files
- Enhanced test output with detailed statistics on XML field usage
- Added error reporting with specific diagnostic messages

**Provides unit tests for Cecille's Namespace parser PR #https://github.com/project-chip/connectedhomeip/pull/37527 and Datatype parser starter PR #[37904](https://github.com/project-chip/connectedhomeip/pull/37904)**

#### Testing
**Platforms:**
- WSL - Windows Subsystem Linux container was used for testing and validating the code in this PR

**Unit Testing Coverage:**
- TestSpecParsingNamespace.py and TestSpecParsingDataType.py are comprehensive unit test modules for validating Matter specification XML files versions 1.4 and 1.4.1, it does some minor checking on 1.3 and master XML's but is not expected to be used for those XML files due to known issues with some of those at the time. Both modules follow a two-phase testing approach: first testing with mock data to verify parser behavior with both valid and invalid inputs, then testing against actual specification XML files. The namespace parser performs extensive validation of namespace structures, ID formats, and version relationships, while maintaining proper error reporting, to the best of my knowledge this covered the requirement for the namespace parser request. While the datatype parser provides deep validation of structs, enums, and bitmaps, including complex field attributes, constraints, and relationships between components, again this is to the best of my knowledge of what was requested for this parser, if anything else needs to be covered I am more than happy to add it. Both modules ensure specification conformance while providing detailed problem reporting for any validation issues encountered.

cecille committed:
DM XMLs 1.4 use alchemy 1.5.4 and add namespaces

Alchemy now does namespace parsing. There was also a change to
the format of the constraints.

cecille committed:
DM XMLs 1.4.1 use alchemy 1.5.4 and add namespaces

cecille committed:
DM XMLs 1.4.1 use alchemy 1.5.4 and add namespaces

cecille committed:
Add namespace parsing to DM script

cecille committed:
Fix build file to add new namespace files

j-ororke committed:
Creating TestSpecParsingNamespace test module:
- Creating with data provided by Cecille in PR here: https://github.com/project-chip/connectedhomeip/pull/37527
- Updated test.yaml to include TestSpecParsingNamespace.py to run in CI
- Updated matter_testing support module to include new NamespacePathLocation function for Namespace file locations
- Updated spec_parsing to include new functions build_xml_namespaces and parse_namespace, as well as XmlNamespaces and XmlTags dataclasses

j-ororke committed:
Updating test_metadata:
- Added TestSpecParsingNamespace.py as it is a unit test and not run against an app.

j-ororke committed:
Updating TestSpecParsingNamespace.py and spec_parsing.py:
- Updating to change method to using 0 instead of 16 bit enums
- Updating to use PrebuiltDatamodelDirectory instead of os to set filepaths
- Updated to utilizing the build_xml_namespaces function in spec_parsing.

j-ororke committed:
Update src/python_testing/matter_testing_infrastructure/chip/testing/spec_parsing.py

Co-authored-by: C Freeman <cecille@google.com>

cecille committed:
DM XMLs: Add parsing for data types [WIP]

Note that anything with a within-data-type conformance will
have a conformance of optional currently. This is OK since we're
not actually using these for anything expect understanding P
markings currently.

TODO: Still not parsing multi-bit bitfields correctly.
TODO: Needs unit tests
TODO: Probably not passing CI on the test support tests
      because of the new errors with the bitfields.

j-ororke committed:
Updating changes for XML spec parsing and adding unit tests:
- Enhanced XmlDataTypeComponent with fields for docs, types, and constraints in spec parsing module
- Added support for detecting optional and nullable fields in XML in spec parsing module
- Improved constraint parsing for min/max values and attribute references in spec parsing module
- Added comprehensive test coverage using 1.4.1 cluster spec XML files in TestSpecParsingDataType unit test module
- Improved test output with detailed statistics on XML field usage in TestSpecParsingDataType unit test module and spec parsing module

j-ororke committed:
Updating TestSpecParsingDataType module:
- Updating some verbiage on comments as needed

j-ororke committed:
Updating test_metadata.yaml:
- Added TestSpecParsingDataType python3 module as it does not run against an app.

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating TestSpecParsingDataType unit testing module:
- Resolving linting errors

j-ororke committed:
Updated TestSpecParsingDataType unit test module:
- Resolving linting errors

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating TestSpecParsingNamespace and spec_parsing modules:
- Resolving linting errors

restyled-commits committed:
Restyled by isort

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating ./scripts/tests/requirements.txt:
- Adding gobject-introspection and libgirepository1.0-dev to resolve issue noticed with girepository-2.0 not being installed when CI checks run.

j-ororke committed:
Updating tests.yaml:
- Added missing dependencies for gobject-introspection libgirepository1.0-dev gir1.2-glib-2.0 in order to get 'girepository-2.0' dependency issue resolved

j-ororke committed:
Merge branch 'datatype_parsing' into create_datatype_parser

j-ororke committed:
Merge branch 'add-namespace-parsing' into create_datatype_parser

j-ororke committed:
Creating Namespace and Datatype parsers using XML and spec_parsing.py code from Cecille's PR #:
Namespace parser:
- Created new TestSpecParsingNamespace.py unit test to test namespaces
- Updated tests.yaml to include TestSpecParsingNamespace.py
- Updated test_metadata.yaml to include TestSpecParsingNamespace.py and reasoning
- Updated matter_testing support module to include new NamespacePathLocation function for Namespace file locations
- Updated spec_parsing to include new functions build_xml_namespaces and parse_namespace, as well as XmlNamespaces and XmlTags dataclasses

Datatype parser:
- spec_parsing module changes:
-- Enhanced XmlDataTypeComponent with fields for docs, types, and constraints
-- Added support for detecting optional and nullable fields in XML
-- Improved constraint parsing for min/max values and attribute references, basically used for dataTypes
-- Improved some error reporting with specific diagnostic messages when parsing dataTypes

- TestSpecParsingDataType unit test module:
-- Added comprehensive test coverage using Matter 1.4.1 cluster XML files
-- Enhanced test output with detailed statistics on XML field usage
-- Added error reporting with specific diagnostic messages

Additionally:
- Removing unneccessary additions in scripts/tests/requirements.txt file
- Adding change to scripts/setup/constraints.txt as was needed to build env

Updates Cecille's PR #37527

andy31415 commented on .github/workflows/tests.yaml:
```diff
@@ -55,6 +55,11 @@ jobs:
+            - name: Install Missing Dependencies
+              run: |
+                apt-get update
+                apt-get install -y gobject-introspection libgirepository1.0-dev gir1.2-glib-2.0
```
Generally we want these as part of whatever base image we need for building/testing.

Could we have a docker image update instead? a apt update & install on every run is not ideal.

j-ororke replied:
Sorry yes, there was an issue noticed with [girepository-2.0 issue](https://csamembers.slack.com/archives/C03JUK7HY6P/p1741599655138339). This code had been moved from a sub-branch with Cecille repo to master in order to file an issue if the issue persisted. However, after moving this code over the issue went away.
So, I should be able to remove the above apt update & apt installs as I don't think they are needed any longer.

andy31415 commented on scripts/setup/constraints.txt:
```diff
@@ -291,4 +291,5 @@ setuptools==68.0.0
 # Higher versions depend on proto-plus, which break
 # nanopb code generation (due to name conflict of the 'proto' module)
-google-api-core==2.17.0
+# google-api-core==2.17.0
```
why do we we have a commented-out line here? why did we downgrade this version?

j-ororke replied:
So, there had been some packages version conflicts noticed to be occurring in the code when building the virtual environment, some reason when I try to run this without doing the downgrade it has been failing with:
- pip._vendor.resolvelib.resolvers.RequirementsConflicted: Requirements conflict

j-ororke replied:
This also appears to have gone away after moving the code here to master. 
I have removed the above dependency version change as it appears no longer needed

andy31415 reviewed (dismissed):
Need to describe/explain the python dependency updates in the summary.
I am not sure what `WSL` means in the testing section.
We want to avoid `apt` commands as part of the CI workflow.

andy31415 commented on src/python_testing/matter_testing_infrastructure/chip/testing/spec_parsing.py:
```diff
@@ -140,6 +172,45 @@ def __str__(self):
+
+
+@dataclass
+class XmlNamespace:
+    """Represents a namespace definition from XML"""
+
+    def __init__(self):
+        self.id: int = 0
+        self.name: str = ""
+        self.tags: dict[int, XmlTag] = {}
+
+    def __str__(self) -> str:
+        tags_str = '\n  '.join(f"{tag_id:04X}: {tag.name}"
+                               for tag_id, tag in sorted(self.tags.items()))
+        return f"Namespace 0x{self.id:04X} ({self.name})\n  {tags_str}"
```
```suggestion
class XmlNamespace:
    """Represents a namespace definition from XML"""
    
    id: int = 0
    name: str = ""
    tags: Dict[int, XmlTag] = field(default_factory=dict)
```

dataclasses auto-create init and str/repr. Do we need a separate str? 
This applies to the rest of the classes as well.

Why are tags a dictonary? are tags themselves not containing an id?

j-ororke replied:
Thank you, I have implemented this in the latest version of the code.

I agree that the additional string formatting function did not appear to be needed, I had just been following the method in the script from device type initiation.

I believe tags do contain an id or at least in the cases I have checked, I believe that the previous reason had just been for additional formatting, also I did not know of the field function previously, so that is super cool and helpful, thank you!

andy31415 commented on src/python_testing/matter_testing_infrastructure/chip/testing/spec_parsing.py:
```diff
@@ -781,11 +978,162 @@ def combine_attributes(base: dict[uint, XmlAttribute], derived: dict[uint, XmlAt
 
 
+"""
+Added XML namespace parsing functions below
```
This should be a comment rather than a string

andy31415 replied:
applies throughout.

j-ororke committed:
Updating tests.yaml, spec_parsing.py, and constraints.txt:
- Updated to remove unneccessary apt install && apt update
- Updated to remove changing version of google-api in constraints.txt as no longer needed
- Updated to change commented out line method as better to have just # for single line entries and removed unneccesary __str__ function to format string before using it.

restyled-commits committed:
Restyled by autopep8

andy31415 commented:
@j-ororke we need to expand "WSL - Windows Subsystem Linux container was used for testing and validating the code in this PR"

What was run and what were the results?

We are asking for detailed text on purpose whenever the text is not "tests are automated and included" to try to encourage people automating tests. However at the same time this PR does include tests, so the text can reference that. Do we have full coverage? or are some things manual (in which case details are needed).

andy31415 commented on scripts/setup/constraints.txt:
```diff
@@ -291,5 +291,5 @@ setuptools==68.0.0
-# google-api-core==2.17.0
-google-api-core==2.11.0
+google-api-core==2.17.0
+# google-api-core==2.11.0
```
still have to remove commented-out code

j-ororke commented:
> @j-ororke we need to expand "WSL - Windows Subsystem Linux container was used for testing and validating the code in this PR"
> 
> What was run and what were the results?
> 
> We are asking for detailed text on purpose whenever the text is not "tests are automated and included" to try to encourage people automating tests. However at the same time this PR does include tests, so the text can reference that. Do we have full coverage? or are some things manual (in which case details are needed).

@andy31415 No problem, I understand the concern here and will be glad to provide some further context. 

**Reason for this PR:**
- Cecille's code contained in this PR creates those XML's to the best of my knowledge and does some of the parsing, we had found an issue that was noticed with CI checks with the original PR's created on Cecille's branch, however the issue is not encountered when combining the parsers together and moving it to master branch, I believe my sub-branches for the parsers on her branch must have gotten corrupted during some of my work and thus is why this PR is now created.

**Testing Coverage:**
- TestSpecParsingNamespace.py and TestSpecParsingDataType.py are comprehensive unit test modules for validating Matter specification XML files versions 1.4 and 1.4.1, it does some minor checking on 1.3 and master XML's but is not expected to be used for those XML files due to known issues with some of those at the time. Both modules follow a two-phase testing approach: first testing with mock data to verify parser behavior with both valid and invalid inputs, then testing against actual specification XML files. The namespace parser performs extensive validation of namespace structures, ID formats, and version relationships, while maintaining proper error reporting, to the best of my knowledge this covered the requirement for the namespace parser request. While the datatype parser provides deep validation of structs, enums, and bitmaps, including complex field attributes, constraints, and relationships between components, again this is to the best of my knowledge of what was requested for this parser, if anything else needs to be covered I am more than happy to add it. Both modules ensure specification conformance while providing detailed problem reporting for any validation issues encountered.

**Notes:**
- To my knowledge the unit tests cover all testing purposes necessary, it is fully automated at this time, there is no reason to include any manual testing to the best of my knowledge. 
- There is no XML's for the 1.3 version for datatype to parse and that there was an issue noticed with the master folder's XML for namespace, in which one XML file ("Namespace-Switches.xml") was noticed to be missing data in a field that was expected to be holding data. However, with that said the parser is only expected to be for the 1.4 and 1.4.1 versions in the /data_model/ folder XML's, not for the 1.3 and master folders XML files as there were known issues with some of those at the time, there are no issues reported with the datatype or namespace parsers for version 1.4 or 1.4.1 data_model XML's.

cecille reviewed (commented):
The namespace PR has landed now, so if you merge master, the namespace file additions will go away and this PR will look a LOT more approachable.

j-ororke committed:
Merge branch 'master' into create_datatype_namespace_parser

j-ororke committed:
Update constraints.txt

Removed commented out line 295 for prior google-api package version

j-ororke committed:
Merge branch 'master' into create_datatype_namespace_parser

j-ororke committed:
Updating TestSpecParsingNamespace and TestSpecParsingDataType python3 test modules:
- Added more comprehensive checks for namespaces between versions to TestSpecParsingNamespace
- Added known current differences between versions for namespaces
- Added change to contain only checks for versions 1.3, 1.4, 1.4.1, and 1.4.2, removed master as noticed in spec_parsing support module

j-ororke committed:
Updating TestSpecParsinNamespace unit test module:
- Minor changes

j-ororke committed:
Updating TestSpecParsingNamespace.py:
- Resolving stylizer issues: isort and autopep8

j-ororke committed:
autopep8 and isort fixes

j-ororke committed:
Merge branch 'master' into create_datatype_namespace_parser

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating TestSpecParsingNamespace as locations of imports have changed recently:
- Added NamespacePathLocation dataclass function to problem_notices support module for correct import
- Changed to importing from problem_notices for NamespacePathLocation, ProblemNotice, ProblemSeverity to be used in TestSpecParsingNamespace module

restyled-commits committed:
Restyled by isort

j-ororke committed:
Resolving linting errors

restyled-commits committed:
Restyled by isort

j-ororke committed:
Merge branch 'master' into create_datatype_namespace_parser

j-ororke committed:
Merge branch 'master' into create_datatype_namespace_parser

j-ororke committed:
Add comprehensive XML validation test for all data types in 1.4.2

Add test_exhaustive_all_fields_in_1_4_2() method to validate every struct,
enum, and bitmap field in all cluster XMLs for version 1.4.2. This test:

- Parses all clusters from /data_model/1.4.2/clusters/
- Validates every struct field has required attributes (name, id, type_info)
- Validates every enum item has required attributes (name, id)
- Validates every bitmap bitfield has required attributes (name, id)
- Skips type_info validation for deprecated fields (e.g., Key field in MonitoringRegistrationStruct)
- Accumulates all issues and reports them at the end instead of failing on first error
- Provides detailed error messages for each validation failure

This ensures full coverage of the data model XMLs and catches any missing
or malformed fields that could affect parsing or downstream analysis.

The test has already identified several clusters with missing type attributes
in ModeOptionStruct fields, which should be addressed in the XML source.

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into create_datatype_namespace_parser
